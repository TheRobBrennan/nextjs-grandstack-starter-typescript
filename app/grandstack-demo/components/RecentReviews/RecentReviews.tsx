import { FC } from "react"
import { useQuery } from "@apollo/react-hooks"
import gql from "graphql-tag"

// Components
import Title from "../Title/Title"

// Material UI
import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
} from "@material-ui/core"

export const GET_RECENT_REVIEWS_QUERY = gql`
  {
    Reviews(options: { sort: [date_DESC] }) {
      date
      user {
        userId
        name
      }
      business {
        name
      }
      text
      stars
    }
  }
`

const RecentReviews: FC = () => {
  const { loading, error, data } = useQuery(GET_RECENT_REVIEWS_QUERY)
  if (loading) return <p>Loading</p>
  if (error) return <p>{error.message}</p>

  return (
    <>
      <Title>Recent Reviews</Title>
      <Table size="small">
        <TableHead>
          <TableRow>
            <TableCell>Date</TableCell>
            <TableCell>Business</TableCell>
            <TableCell>User</TableCell>
            <TableCell>Review</TableCell>
            <TableCell align="right">Stars</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {data?.Reviews.map((review) => (
            <TableRow
              key={`${review?.user?.name}-${review?.user?.userId}-${review?.date?.year?.low}-${review?.date?.month?.low}-${review?.date?.day?.low}`}
            >
              <TableCell>
                {`${review?.date?.year?.low}.${review?.date?.month?.low}.${review?.date?.day?.low}`}
              </TableCell>
              <TableCell>{review?.business?.name}</TableCell>
              <TableCell>{review?.user?.name}</TableCell>
              <TableCell>{review?.text}</TableCell>
              <TableCell align="right">{review?.stars?.low}</TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </>
  )
}
export default RecentReviews
