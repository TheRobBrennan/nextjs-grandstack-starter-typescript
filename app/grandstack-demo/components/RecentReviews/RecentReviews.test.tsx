import React from "react"
import TestRenderer, { act } from "react-test-renderer"
import { MockedProvider } from "@apollo/client/testing"

import RecentReviews, { GET_RECENT_REVIEWS_QUERY } from "./RecentReviews"

describe("GRANDstack RecentReviews component", () => {
  it("should render the latest reviews after receiving data", async () => {
    // Define our Apollo request
    const renderRequest = {
      request: {
        query: GET_RECENT_REVIEWS_QUERY,
        variables: {},
      },
      result: {
        data: {
          Reviews: [
            {
              date: {
                year: {
                  low: 2018,
                  high: 0,
                },
                month: {
                  low: 9,
                  high: 0,
                },
                day: {
                  low: 10,
                  high: 0,
                },
              },
              user: {
                userId: "u4",
                name: "Angie",
                __typename: "User",
              },
              business: {
                name: "Imagine Nation Brewing",
                __typename: "Business",
              },
              text: "",
              stars: {
                low: 3,
                high: 0,
              },
              __typename: "Review",
            },
            {
              date: {
                year: {
                  low: 2018,
                  high: 0,
                },
                month: {
                  low: 8,
                  high: 0,
                },
                day: {
                  low: 11,
                  high: 0,
                },
              },
              user: {
                userId: "u4",
                name: "Angie",
                __typename: "User",
              },
              business: {
                name: "Zootown Brew",
                __typename: "Business",
              },
              text: "",
              stars: {
                low: 5,
                high: 0,
              },
              __typename: "Review",
            },
            {
              date: {
                year: {
                  low: 2018,
                  high: 0,
                },
                month: {
                  low: 3,
                  high: 0,
                },
                day: {
                  low: 24,
                  high: 0,
                },
              },
              user: {
                userId: "u2",
                name: "Bob",
                __typename: "User",
              },
              business: {
                name: "Market on Front",
                __typename: "Business",
              },
              text: "",
              stars: {
                low: 4,
                high: 0,
              },
              __typename: "Review",
            },
            {
              date: {
                year: {
                  low: 2018,
                  high: 0,
                },
                month: {
                  low: 1,
                  high: 0,
                },
                day: {
                  low: 3,
                  high: 0,
                },
              },
              user: {
                userId: "u1",
                name: "Will",
                __typename: "User",
              },
              business: {
                name: "Ninja Mike's",
                __typename: "Business",
              },
              text:
                "Best breakfast sandwich at the Farmer's Market. Always get the works.",
              stars: {
                low: 4,
                high: 0,
              },
              __typename: "Review",
            },
            {
              date: {
                year: {
                  low: 2017,
                  high: 0,
                },
                month: {
                  low: 11,
                  high: 0,
                },
                day: {
                  low: 13,
                  high: 0,
                },
              },
              user: {
                userId: "u3",
                name: "Jenny",
                __typename: "User",
              },
              business: {
                name: "Ninja Mike's",
                __typename: "Business",
              },
              text: "",
              stars: {
                low: 5,
                high: 0,
              },
              __typename: "Review",
            },
            {
              date: {
                year: {
                  low: 2016,
                  high: 0,
                },
                month: {
                  low: 11,
                  high: 0,
                },
                day: {
                  low: 21,
                  high: 0,
                },
              },
              user: {
                userId: "u3",
                name: "Jenny",
                __typename: "User",
              },
              business: {
                name: "Hanabi",
                __typename: "Business",
              },
              text: "",
              stars: {
                low: 5,
                high: 0,
              },
              __typename: "Review",
            },
            {
              date: {
                year: {
                  low: 2016,
                  high: 0,
                },
                month: {
                  low: 7,
                  high: 0,
                },
                day: {
                  low: 14,
                  high: 0,
                },
              },
              user: {
                userId: "u3",
                name: "Jenny",
                __typename: "User",
              },
              business: {
                name: "KettleHouse Brewing Co.",
                __typename: "Business",
              },
              text: "",
              stars: {
                low: 5,
                high: 0,
              },
              __typename: "Review",
            },
            {
              date: {
                year: {
                  low: 2016,
                  high: 0,
                },
                month: {
                  low: 3,
                  high: 0,
                },
                day: {
                  low: 4,
                  high: 0,
                },
              },
              user: {
                userId: "u1",
                name: "Will",
                __typename: "User",
              },
              business: {
                name: "Ducky's Car Wash",
                __typename: "Business",
              },
              text: "Awesome full-service car wash. Love Ducky's!",
              stars: {
                low: 5,
                high: 0,
              },
              __typename: "Review",
            },
            {
              date: {
                year: {
                  low: 2016,
                  high: 0,
                },
                month: {
                  low: 1,
                  high: 0,
                },
                day: {
                  low: 3,
                  high: 0,
                },
              },
              user: {
                userId: "u1",
                name: "Will",
                __typename: "User",
              },
              business: {
                name: "KettleHouse Brewing Co.",
                __typename: "Business",
              },
              text: "Great IPA selection!",
              stars: {
                low: 4,
                high: 0,
              },
              __typename: "Review",
            },
            {
              date: {
                year: {
                  low: 2015,
                  high: 0,
                },
                month: {
                  low: 12,
                  high: 0,
                },
                day: {
                  low: 15,
                  high: 0,
                },
              },
              user: {
                userId: "u2",
                name: "Bob",
                __typename: "User",
              },
              business: {
                name: "Imagine Nation Brewing",
                __typename: "Business",
              },
              text: "",
              stars: {
                low: 4,
                high: 0,
              },
              __typename: "Review",
            },
            {
              date: {
                year: {
                  low: 2015,
                  high: 0,
                },
                month: {
                  low: 9,
                  high: 0,
                },
                day: {
                  low: 1,
                  high: 0,
                },
              },
              user: {
                userId: "u1",
                name: "Will",
                __typename: "User",
              },
              business: {
                name: "Neo4j",
                __typename: "Business",
              },
              text: "The world's leading graph database HQ!",
              stars: {
                low: 5,
                high: 0,
              },
              __typename: "Review",
            },
            {
              date: {
                year: {
                  low: 2015,
                  high: 0,
                },
                month: {
                  low: 8,
                  high: 0,
                },
                day: {
                  low: 29,
                  high: 0,
                },
              },
              user: {
                userId: "u1",
                name: "Will",
                __typename: "User",
              },
              business: {
                name: "Missoula Public Library",
                __typename: "Business",
              },
              text:
                "Not a great selection of books, but fortunately the inter-library loan system is good. Wifi is quite slow. Not many comfortable places to site and read. Looking forward to the new building across the street in 2020!",
              stars: {
                low: 3,
                high: 0,
              },
              __typename: "Review",
            },
          ],
        },
      },
    }

    // Define our mock response(s)
    const gqlMocks = [renderRequest]

    // Verify success state
    const component = TestRenderer.create(
      <MockedProvider mocks={gqlMocks} addTypename={false}>
        <RecentReviews />
      </MockedProvider>
    )

    // Advance to the next tick in the event loop so our chart can render
    await act(() => {
      return new Promise((resolve) => {
        setTimeout(resolve, 0)
      })
    })

    // Verify
    const p = component.root.findByType("h2")
    expect(p.children).toContain("Recent Reviews")
  })

  describe("should display an error message if", () => {
    it("there is a network request failure", async () => {
      // Define our Apollo request
      const renderRequest = {
        request: {
          query: GET_RECENT_REVIEWS_QUERY,
          variables: {},
        },
        error: new Error("Simulated network or HTTP error"),
      }

      // Define our mock response(s)
      const gqlMocks = [renderRequest]

      // Verify success state
      const component = TestRenderer.create(
        <MockedProvider mocks={gqlMocks} addTypename={false}>
          <RecentReviews />
        </MockedProvider>
      )

      // Advance to the next tick in the event loop so our chart can render
      await act(() => {
        return new Promise((resolve) => {
          setTimeout(resolve, 0)
        })
      })

      // Verify
      const tree = component.toJSON()
      expect(tree.children).toContain("Simulated network or HTTP error")
    })
    it("our GraphQL server successfully processed our request and has returned with one or more error(s) in the result", async () => {
      // Define our Apollo request
      const renderRequest = {
        request: {
          query: GET_RECENT_REVIEWS_QUERY,
          variables: {},
        },
        result: {
          // REVISIT: Create a pull request - errors should be able to be defined on the GraphQL result object because it can contain data and errors
          //          https://www.apollographql.com/docs/react/development-testing/testing/#graphql-errors
          errors: [
            new Error("Simulated GraphQL server response with an error"),
          ],
        },
      }

      // Define our mock response(s)
      const gqlMocks = [renderRequest]

      // Verify success state
      const component = TestRenderer.create(
        <MockedProvider mocks={gqlMocks} addTypename={false}>
          <RecentReviews />
        </MockedProvider>
      )

      // Advance to the next tick in the event loop so our chart can render
      await act(() => {
        return new Promise((resolve) => {
          setTimeout(resolve, 0)
        })
      })

      // Verify
      const tree = component.toJSON()
      expect(tree.children).toContain(
        "Simulated GraphQL server response with an error"
      )
    })
  })
})
