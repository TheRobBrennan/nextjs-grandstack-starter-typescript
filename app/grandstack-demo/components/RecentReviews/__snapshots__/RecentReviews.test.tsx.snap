// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GRANDstack RecentReviews component should render the latest reviews after receiving data 1`] = `"<h2 class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom\\">Recent Reviews</h2><table class=\\"MuiTable-root\\"><thead class=\\"MuiTableHead-root\\"><tr class=\\"MuiTableRow-root MuiTableRow-head\\"><th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall\\" scope=\\"col\\">Date</th><th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall\\" scope=\\"col\\">Business</th><th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall\\" scope=\\"col\\">User</th><th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall\\" scope=\\"col\\">Review</th><th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall\\" scope=\\"col\\">Stars</th></tr></thead><tbody class=\\"MuiTableBody-root\\"><tr class=\\"MuiTableRow-root\\"><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">2018.9.10</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Imagine Nation Brewing</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Angie</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\"></td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall\\">3</td></tr><tr class=\\"MuiTableRow-root\\"><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">2018.8.11</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Zootown Brew</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Angie</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\"></td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall\\">5</td></tr><tr class=\\"MuiTableRow-root\\"><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">2018.3.24</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Market on Front</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Bob</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\"></td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall\\">4</td></tr><tr class=\\"MuiTableRow-root\\"><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">2018.1.3</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Ninja Mike's</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Will</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Best breakfast sandwich at the Farmer's Market. Always get the works.</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall\\">4</td></tr><tr class=\\"MuiTableRow-root\\"><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">2017.11.13</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Ninja Mike's</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Jenny</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\"></td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall\\">5</td></tr><tr class=\\"MuiTableRow-root\\"><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">2016.11.21</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Hanabi</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Jenny</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\"></td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall\\">5</td></tr><tr class=\\"MuiTableRow-root\\"><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">2016.7.14</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">KettleHouse Brewing Co.</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Jenny</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\"></td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall\\">5</td></tr><tr class=\\"MuiTableRow-root\\"><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">2016.3.4</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Ducky's Car Wash</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Will</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Awesome full-service car wash. Love Ducky's!</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall\\">5</td></tr><tr class=\\"MuiTableRow-root\\"><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">2016.1.3</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">KettleHouse Brewing Co.</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Will</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Great IPA selection!</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall\\">4</td></tr><tr class=\\"MuiTableRow-root\\"><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">2015.12.15</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Imagine Nation Brewing</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Bob</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\"></td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall\\">4</td></tr><tr class=\\"MuiTableRow-root\\"><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">2015.9.1</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Neo4j</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Will</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">The world's leading graph database HQ!</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall\\">5</td></tr><tr class=\\"MuiTableRow-root\\"><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">2015.8.29</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Missoula Public Library</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Will</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Not a great selection of books, but fortunately the inter-library loan system is good. Wifi is quite slow. Not many comfortable places to site and read. Looking forward to the new building across the street in 2020!</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall\\">3</td></tr></tbody></table>"`;
